packages/@react-stately/table/src/useTreeGridState.ts

```diff
@@ -0,0 +1,314 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import {CollectionBuilder} from '@react-stately/collections';
+import {Expandable} from '@react-types/shared';
+import {GridNode} from '@react-types/grid';
+import {Key, useMemo} from 'react';
+import {TableCollection} from './TableCollection';
+import {TableState, TableStateProps, useTableState} from './useTableState';
+import {useControlledState} from '@react-stately/utils';
+
+export interface TreeGridState<T> extends TableState<T> {
+  /** A set of keys for items that are expanded. */
+  expandedKeys: 'all' | Set<Key>,
+  /** Toggles the expanded state for a row by its key. */
+  toggleKey(key: Key): void,
+  /** The key map containing nodes representing the collection's tree grid structure. */
+  keyMap: Map<Key, GridNode<T>>,
+  /** The number of columns provided by the user. */
+  userColumnCount: number
+}
+
+// TODO: For now I am omitting collection since this is a bit of a special case where are still using TableCollection but include
+// extra information for the tree grid in the state. That means the user can't really provide their own collection. Perhaps it would
+// be good to still have a TreeGridCollection implementation in that case?
+export interface TreeGridStateProps<T> extends Expandable, Omit<TableStateProps<T>, 'collection'> {}
+
+/**
+ * Provides state management for a tree grid component. Handles building a collection
+ * of columns and rows from props. In addition, it tracks and manages expanded rows, row selection, and sort order changes.
+ */
+export function useTreeGridState<T extends object>(props: TreeGridStateProps<T>): TreeGridState<T> {
+  let {
+    selectionMode = 'none',
+    showSelectionCheckboxes,
+    showDragButtons,
+    expandedKeys: propExpandedKeys,
+    defaultExpandedKeys: propDefaultExpandedKeys,
+    onExpandedChange,
+    children
+  } = props;
+
+  let [expandedKeys, setExpandedKeys] = useControlledState(
+    propExpandedKeys ? convertExpanded(propExpandedKeys) : undefined,
+    propDefaultExpandedKeys ? convertExpanded(propDefaultExpandedKeys) : new Set(),
+    onExpandedChange
+  );
+
+  let context = useMemo(() => ({
+    showSelectionCheckboxes: showSelectionCheckboxes && selectionMode !== 'none',
+    showDragButtons: showDragButtons,
+    selectionMode,
+    columns: []
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  }), [children, showSelectionCheckboxes, selectionMode, showDragButtons]);
+
+  let builder = useMemo(() => new CollectionBuilder<T>(), []);
+  // TODO: figure out this ts issue
+  // @ts-ignore
+  let nodes = useMemo(() => builder.build({children}, context), [builder, children, context]);
```

Comment:
TS issue here since children of CollectionBuilder isn't strictly compatible with a TableBody child. I could expand the definition of CollectionElement to `export type CollectionElement<T> = SectionElement<T> | ItemElement<T> | ReactElement<TableBodyProps<T>>;` but that brings in table specific details to the general collection types... Frankly not sure how useTableState doesn't run into this issue

Reply:
Why can't you use `useCollection` here rather than CollectionBuilder directly? It looks like that accepts a more general type. I guess the factory function could call generateTreeGridCollection and then construct a TableCollection?

Reply:
The only issue would be extracting the `keyMap` and `columnCount` from the `generateTreeGridCollection` call since it would be in the factory function. I guess I could set up a ref or something to store those

Reply:
Oh hmm. It's fine if it doesn't work, a ref doesn't sound great either. Maybe the types can help though.

Reply:
tried refactoring this a bit, ran into an infinite rerender when including the generateTreeGridCollection call into the factory function useCallback along with a state setter to store the treeGridCollection:
```
  let [treeGridCollection, setTreeGridCollection] = useState<TreeGridCollection<T>>({keyMap: new Map(), tableNodes: [], flattenedRows: [], userColumnCount: 0});
  let factory = useCallback((nodes) => {
    let treeGridCollection = generateTreeGridCollection<T>(nodes, {showSelectionCheckboxes, showDragButtons, expandedKeys});
    setTreeGridCollection(treeGridCollection);
    return new TableCollection(treeGridCollection.tableNodes, null, context);
  }, [context, showSelectionCheckboxes, showDragButtons, expandedKeys]);

  let collection = useCollection<T, ITableCollection<T>>({children}, factory, context);
```
gonna punt for a bit and handle the other todos before trying to figure out why

Reply:
pushed a type cast that gets rid of the `ts-ignore`, feels a bit scuffed